circle.yml

```diff
@@ -1,13 +1,22 @@
 machine:
+  environment:
+    YARN_VERSION: 0.18.1
+    PATH: "${PATH}:${HOME}/.yarn/bin:${HOME}/${CIRCLE_PROJECT_REPONAME}/node_modules/.bin"
   node:
     version: 5.5.0
 
-  post:
-    - curl -o- -L https://yarnpkg.com/install.sh | bash
-
 dependencies:
+  pre:
+    - |
+      if [[ ! -e ~/.yarn/bin/yarn || $(yarn --version) != "${YARN_VERSION}" ]]; then
+        echo "Download and install Yarn."
+        curl -o- -L https://yarnpkg.com/install.sh | bash -s -- --version $YARN_VERSION
```

Comment:
If yarn is installed with a bad version and we try to install another version, we will have the same problem : 
`mkdir: cannot create directory ‘.yarn’: File exists`

I think you have to uninstall the previous version of yarn or delete `.yarn` file (can be enough)